### DISCUSSION TITLE
DRB3 functionality support

### DISCUSSION THREAD

**Discussion by darkcorpd** [laszlodaniel/ChryslerScanner#18] [General]
Hi Daniel!

I've tried to execute the Fuel Sync procedure with the Chrysler Scanner but it won't start if some conditions aren't met.
For example, scanner shows engine temperature 26-28 Celsius degrees while actually my engine is warmed up and other elm tools and dashboard show around 90-95 degrees.
Is there any conversation problem or I've missed something (like scanner temperature is something like an average of engine, intake air, battery and ambient temperatures)?
![IMG20241106185047~01](https://github.com/user-attachments/assets/070a24c3-e9f0-4764-968d-8df273166c9d)

**Comment by laszlodaniel** [comment]
Hi Albert, it could be that 14-05 is not ECT anymore in later model years, but some other temperature.
Can you share the part number of this PCM calibration? I'll investigate.
An SCI-bus log of the failed attempt would also be helpful.

**Reply by darkcorpd** [reply]
My ECU p/n is 56041423AH.
This is 1999 Jeep Grand Cherokee WJ 4.0 with Federal emissions.

**Reply by laszlodaniel** [reply]
Verified in multiple JTEC+ flash files that SCI 14 05 returns ECT and is present in CCD 0C as well.
However CCD 8C, which is usually used to broadcast ECT and IAT (intake air temperature) in SBEC3, uses different memory values for temperature.
Why this resembles ambient temperature is beyond me.
JTEC+ is biting my ass again...
ELM327 is reading ECT via generic OBD2 request, so it's not the same as SCI 14 05.

What error do you get when you start Set Fuel Sync?
The GUI only checks if engine is running, then the other commands should go through.
Details needed.

**Comment by laszlodaniel** [comment]
New Android application is out with Set fuel sync support!

[DRB Pro Emulator](https://mega.nz/file/UCEhibgT#bNUTWvdelKZIlwPj7Y6pUJl9RV_p2pmTZkSOA7xrUFU)

<img src="https://chryslerccdsci.wordpress.com/wp-content/uploads/2025/01/screenshot_20250130_130351.jpg" alt="Screenshot" width="300">

<img src="https://chryslerccdsci.wordpress.com/wp-content/uploads/2025/01/screenshot_20250130_130405.jpg" alt="Screenshot" width="300">

<img src="https://chryslerccdsci.wordpress.com/wp-content/uploads/2025/01/screenshot_20250130_130416.jpg" alt="Screenshot" width="300">

<img src="https://chryslerccdsci.wordpress.com/wp-content/uploads/2025/01/screenshot_20250130_130426.jpg" alt="Screenshot" width="300">


This can be found under the "Diagnostics / Engine / Miscellaneous / Set fuel sync" menu.

The parameter IDs are not properly resolved yet, so the coolant temperature may still look wrong to you.
You can disable some preconditions if they do not seem right.

**Reply by darkcorpd** [reply]
That' great news!
The application asks for login/password. Where can I find them?

**Reply by laszlodaniel** [reply]
You have to sign up with the "SIGN UP" button for a new account.
You will receive a verification E-mail afterwards, and some free tokens for later use.

**Reply by darkcorpd** [reply]
Thanks, have signed up and verified email successfully. But on Home screen the notification that email is not verified is still active although on Profile page the email status is Verified.
BTW how to connect to Scanner via Bluetooth? I see CryPro2 device but connection attemps are unsuccessful.

**Reply by laszlodaniel** [reply]
Yes, that's a bug. Sign out and sign in again, the note will disappear.

Make sure that the scanner firmware is updated to latest version and Bluetooth permission is granted.
Permission is asked on first connection attempt.

**Reply by darkcorpd** [reply]
Updated the firmware to the latest, ok
Scanner connects via Bluetooth, ok.

Unfortunately, fuel synchronisation doesn't work for me yet. 

It fails on prepositions even though the engine is started and warmed up to operating temperature. Nothing special happens when pressing the Start button.
![Screenshot_2025-02-01-10-32-35-63_8468db5ea37d174f590ca6ecff2be1aa](https://github.com/user-attachments/assets/a9f6b13c-b44a-488b-893b-6d6ae9ad1c76)
![Screenshot_2025-02-01-10-49-45-12_8468db5ea37d174f590ca6ecff2be1aa](https://github.com/user-attachments/assets/cb1337be-5a7e-44c8-b4cf-bff74b02f76f)


https://github.com/user-attachments/assets/3fb60a8f-4d8e-4be8-9e23-5ff82486dff7

**Reply by laszlodaniel** [reply]
Thanks for the tests! The engine speed is checked once again before starting, and it's reading the rpm from an invalid region of RAM. This is why syncing never starts...
Clearly your PCM works differently than those I run the tests on.
I'll figure out a solution next week.

Is this still the 56041423AH 1999 Jeep Grand Cherokee WJ 4.0 with Federal emissions PCM?

**Reply by darkcorpd** [reply]
Yup

**Reply by darkcorpd** [reply]
Installed the latest Android app update (v0.4.6).
Fuel sync task starts regardless of preconditions.
Actually it only sets revs to 1000 RPM and nothing happens.
No change in Basic ignitoin timing field or main Fuel sync degree field.

![WhatsApp Image 2025-02-07 at 11 06 39](https://github.com/user-attachments/assets/b2df3bec-d78a-4535-b3a8-b19a7350f092)

**Reply by laszlodaniel** [reply]
Ok, thanks for testing!

I took another deep dive into JTEC/JTEC+ firmwares and found an explanation why this is still not working out.

Here is the expected workflow:

```
1. Enable basic ignition timing:

TX: 21 01    <- enable basic ignition timing
RX: 21 01 01 <- basic ignition timing enabled
RX: 21 01 02 <- error: throttle open
RX: 21 01 03 <- error: transmission in drive

2. Raise engine idle speed:

TX: 21 10    <- raise engine idle speed to preset value (~1000 RPM)
RX: 21 10 10 <- engine idle speed raised

3. Request fuel sync calculation update and read set fuel sync (on repeat, every 200 ms or so):

TX: 21 10    <- request calculation (any non-zero parameter is ok)
RX: 21 10    <- no additional response
TX: 14 2A    <- read set fuel sync
RX: 14 2A XX <- XX = set fuel sync

4. Finish:

TX: 21 00    <- disable basic ignition timing and return to normal idle speed
RX: 21 00 00 <- ok
```

Can you try these steps with the Chrysler Scanner GUI?
You can send these commands and setup message repeat from the SCI-bus tab.
If it works I implement it in the Android app as well.

Interestingly JTEC PCMs don't need separate "SCI 19 XX" commands to raise idle speed, it is done automatically when sending the first "SCI 21 10" command. Sending this a second time triggers the PCM to update the set sync value, which can be read with "SCI 14 2A".

Disassembled and reversed code for reference:

```
ROM:D338 SCI_IGNTIM:                             ; DATA XREF: ROM:F05C↓o
ROM:D338                 ldab    348h, Z         ; B = payload | SCI 21 XX
ROM:D33C                 beq     Finish          ; B = 0 | SCI 21 00 - finish setting fuel sync
ROM:D33E
ROM:D33E loc_D33E:                               ; CODE XREF: sub_140F0+74↓p
ROM:D33E                 brclr   0D6h, Z, #80h, ThrottleClosed
ROM:D344                 ldab    #2              ; B = 2 | error: throttle open
ROM:D346                 bra     Finish
ROM:D348 ; ---------------------------------------------------------------------------
ROM:D348
ROM:D348 ThrottleClosed:                         ; CODE XREF: SCI_SETACT:loc_D33E↑j
ROM:D348                 brset   CFG_TRSTYP, #80h, TransOk ; manual transmission
ROM:D34E                 brset   10Bh, Z, #40h, TransOk
ROM:D354                 ldab    #3              ; B = 3 | error: transmission in drive
ROM:D356                 bra     Finish
ROM:D358 ; ---------------------------------------------------------------------------
ROM:D358
ROM:D358 TransOk:                                ; CODE XREF: SCI_SETACT:ThrottleClosed↑j
ROM:D358                                         ; SCI_SETACT+412↑j
ROM:D358                 brset   109h, Z, #40h, CalcFuelSync
ROM:D35E                 cmpb    #10h            ; B = 10 | SCI 21 10 - raise idle speed
ROM:D360                 beq     RaiseIdleSpeed
ROM:D362                 cmpb    #1              ; B = 1 | SCI 21 01 - enable basic ignition timing
ROM:D364                 bne     Finish
ROM:D366                 bset    109h, Z, #80h
ROM:D36A                 bra     Response
ROM:D36C ; ---------------------------------------------------------------------------
ROM:D36C
ROM:D36C RaiseIdleSpeed:                         ; CODE XREF: SCI_SETACT+424↑j
ROM:D36C                 bset    109h, Z, #40h
ROM:D370                 lde     RPM_IDLSFS      ; Engine idle speed for setting fuel sync
ROM:D374                 ste     RAM_IDLTRG, Z   ; Target engine idle speed (1 bit = 0.125 rpm)
ROM:D378
ROM:D378 Response:                               ; CODE XREF: SCI_SETACT+42E↑j
ROM:D378                 jsr     SCI_TX          ; Echo B to SCI-bus
ROM:D37C                 jmp     SCI_ATMEND
ROM:D380 ; ---------------------------------------------------------------------------
ROM:D380
ROM:D380 Finish:                                 ; CODE XREF: SCI_SETACT+400↑j
ROM:D380                                         ; SCI_SETACT+40A↑j ...
ROM:D380                 jsr     SCI_TX          ; Echo B to SCI-bus
ROM:D384                 bclr    109h, Z, #0C0h  ; clear status bits (0x80 and 0x40)
ROM:D388                 jmp     SCI_ATMEND
ROM:D38C ; ---------------------------------------------------------------------------
ROM:D38C
ROM:D38C CalcFuelSync:                           ; CODE XREF: SCI_SETACT:TransOk↑j
ROM:D38C                 ldab    #0Fh
ROM:D38E                 tbxk
ROM:D390                 ldx     #80E5h
ROM:D394                 clrb
ROM:D396
ROM:D396 Loop:                                   ; CODE XREF: SCI_SETACT+466↓j
ROM:D396                 tst     0, X
ROM:D398                 beq     ExitLoop
ROM:D39A                 aix     #5
ROM:D39C                 incb
ROM:D39E                 cmpb    #4
ROM:D3A0                 bcc     locret_D3F6
ROM:D3A2                 bra     Loop
ROM:D3A4 ; ---------------------------------------------------------------------------
ROM:D3A4
ROM:D3A4 ExitLoop:                               ; CODE XREF: SCI_SETACT+45C↑j
ROM:D3A4                 lde     1, X
ROM:D3A8                 ldd     3, X
ROM:D3AA                 subd    8, X
ROM:D3AC                 sbce    6, X
ROM:D3B0                 ande    #0FFFh
ROM:D3B4                 lsre
ROM:D3B6                 rord
ROM:D3B8                 lsre
ROM:D3BA                 rord
ROM:D3BC                 lsre
ROM:D3BE                 rord
ROM:D3C0                 xgdx
ROM:D3C2                 ldd     14Ch, Z
ROM:D3C6                 fdiv
ROM:D3C8                 xgdx
ROM:D3CA                 brset   byte_2707, #20h, loc_D3D6
ROM:D3D0                 ldab    byte_8D9
ROM:D3D4                 bra     loc_D3DA
ROM:D3D6 ; ---------------------------------------------------------------------------
ROM:D3D6
ROM:D3D6 loc_D3D6:                               ; CODE XREF: SCI_SETACT+48E↑j
ROM:D3D6                 ldab    byte_8D8
ROM:D3DA
ROM:D3DA loc_D3DA:                               ; CODE XREF: SCI_SETACT+498↑j
ROM:D3DA                 mul
ROM:D3DC                 adca    #0
ROM:D3DE                 brset   byte_2707, #20h, loc_D3EA
ROM:D3E4                 ldab    byte_8D9
ROM:D3E8                 bra     SaveFuelSync
ROM:D3EA ; ---------------------------------------------------------------------------
ROM:D3EA
ROM:D3EA loc_D3EA:                               ; CODE XREF: SCI_SETACT+4A2↑j
ROM:D3EA                 ldab    byte_8D8
ROM:D3EE
ROM:D3EE SaveFuelSync:                           ; CODE XREF: SCI_SETACT+4AC↑j
ROM:D3EE                 lsrb
ROM:D3F0                 sba
ROM:D3F2                 staa    35Bh, Z         ; SCI 14 2A
ROM:D3F6
ROM:D3F6 locret_D3F6:                            ; CODE XREF: SCI_SETACT+464↑j
ROM:D3F6                 rts
```

**Reply by luizrenault** [reply]
Hi. Can you help out on the steps to save and disassembly the firmware for a 1997 Cherokee XJ JTEC?

I bought a reader and had some trouble with the android app.

**Reply by laszlodaniel** [reply]
Hi! Yes, I've just started a new thread focused on JTEC firmware disassembly here:
https://github.com/laszlodaniel/ChryslerScanner/discussions/23

I'll also be uploading a YouTube tutorial tomorrow that walks through how to read and save the JTEC flash using the Android app step by step. I'll post the link as soon as it's ready!

**Reply by darkcorpd** [reply]
Hi Daniel!
Finally I've successfully done fuel sync by Android app.
Although it doesn't correctly show required parameters (etc. no RPM, wrong temperature) but if you skip preparation then revs will jump from idle to 1000 RPM and during those leaps the App shows the actual degree.
Mine was -2, I've set it to +4 as recommended above.

Note: actual degree moves 4-5 DEG backwards while you tighten the fixing screw, so it needs to be set with adding some extra compensative value. For ex. I've set to 8 DEG, after tightening it shows desired 4 DEG.

**Reply by laszlodaniel** [reply]
This is wonderful news, Albert! Thanks for letting me know.

I've spent the past few months building a new foundation for diagnostics, allowing sensor data to be requested directly from the recognized PCM. As you know, the current sensor PIDs are hardcoded for a specific SBEC3 unit. In future updates, the Android app will identify the detected module from the DRB3 database, along with all its sensors, actuator tests, and more (or fall back to a generic map for that controller). This will make it possible to display the correct sensors throughout the app.

Your observation about the sync degrees is very helpful! I remember reading an instruction that said not to loosen the bolt completely, but just enough to maintain some friction. It's smart thinking to read the sync again for verification.
